File: AdventOfCode2022/Python/Day10/Day10.py
def get_inputs(file):
    """
    function to read in data from a txt file
    """

    with open(file) as f:
        data = f.read().split('\n')

    return data

def Part1(filename):
    """
    Solutions to Part 1
    """
    data = get_inputs(filename)

    cycle_number = 1
    X = 1
    key_cycles = []

    for entry in data:
        instruction = entry.split(' ')[0]
        
        if instruction == 'noop':

            cycle_number += 1
            for key_cycle in [20, 60, 100, 140, 180, 220]:
                if cycle_number == key_cycle:
                    key_cycles.append(X*cycle_number)
            
        if instruction == 'addx':

            value = int(entry.split(' ')[1])

            cycle_number += 1
            for key_cycle in [20, 60, 100, 140, 180, 220]:
                if cycle_number == key_cycle:
                    key_cycles.append(X*cycle_number)
            
            X += value
            
            cycle_number += 1
            for key_cycle in [20, 60, 100, 140, 180, 220]:
                if cycle_number == key_cycle:
                    key_cycles.append(X*cycle_number)
    
    return sum(key_cycles)

def Part2(filename):
    """
    Solutions to Part 2
    """
    data = get_inputs(filename)
    
    cycle_number = 1
    X = 1
    pixels = list('.'*40*6)

    for entry in data:

        instruction = entry.split(' ')[0]

        if instruction == 'noop':
            
            for pos in [X-1, X, X+1]:
                for row in [0,1,2,3,4,5]:
                   if ((40*row)+1 <= cycle_number <= 40*(row+1)) and (cycle_number-1 - (row*40) == pos):
                        pixels[cycle_number-1] = '#'
                
            cycle_number += 1

        if instruction == 'addx':

            value = int(entry.split(' ')[1])

            for pos in [X-1, X, X+1]:
                for row in [0,1,2,3,4,5]:
                    if ((40*row)+1 <= cycle_number <= 40*(row+1)) and (cycle_number-1 - (row*40) == pos):
                        pixels[cycle_number-1] = '#'

            cycle_number += 1

            for pos in [X-1, X, X+1]:
                for row in [0,1,2,3,4,5]:
                    if ((40*row)+1 <= cycle_number <= 40*(row+1)) and (cycle_number-1 - (row*40) == pos):
                        pixels[cycle_number-1] = '#'

            cycle_number += 1
            X += value
                
    str_ = ''
    for pixel in pixels:
        str_ += pixel

    print('.'+'-'*41+'.')
    for i in range(0,6):
        print('| '+str_[i*40:(i+1)*40-1] +' |')
    print('.'+'-'*41+'.')
    
    return "read above output"

def main():
    """
    main function
    """ 
    testfile = "tests.txt"
    inputfile = "inputs.txt"

    print("Advent-Of-Code 2022 - Day10")

    print(f'"{testfile}": Answer to Part 1 = {Part1(testfile)}')
    print(f'"{testfile}": Answer to Part 2 = {Part2(testfile)}')

    print(f'"{inputfile}": Answer to Part 1 = {Part1(inputfile)}')
    print(f'"{inputfile}": Answer to Part 2 = {Part2(inputfile)}')

if __name__ == "__main__":
    main()

File: AdventOfCode2022/Go/Day10/Day10.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func Parse(filename string) []string {

	file, _ := os.ReadFile(filename)

	data := strings.Split(string(file), "\n")

	return data
}

func Part1(filename string) int {

	instructions := Parse(filename)

	sums := 0
	cycle := 0
	x := 1

	for _, instruction := range instructions {

		cycle++

		if (cycle-20)%40 == 0 {
			sums += x * cycle
		}

		if instruction[:4] != "noop" {

			cycle++

			if (cycle-20)%40 == 0 {
				sums += x * cycle
			}

			xp, _ := strconv.Atoi((strings.TrimSpace(instruction[5:])))
			x += xp
		}
	}
	return sums
}

func Part2(filename string, printing bool) int {

	// create blank canvas
	pixels := []string{}
	for i := 0; i <= 40*6; i++ {
		pixels = append(pixels, ".")
	}

	instructions := Parse(filename)

	cycle := 1
	x := 1

	for _, instruction := range instructions {

		if printing {
			fmt.Print("Start cycle   ", cycle, ": begin executing ", instruction, "\n")
		}

		if x+(40*int(math.Floor((float64(cycle)/40)))) <= cycle && cycle <= x+2+(40*int(math.Floor((float64(cycle)/40)))) {
			pixels[cycle-1] = "#"
		}

		if printing {
			fmt.Print("During cycle  ", cycle, ": CRT draws pixel in position ", cycle-1, "\n")
			fmt.Print("Current CRT row: ", pixels[:cycle], "\n\n")
		}

		// AddX
		if instruction[:4] != "noop" {

			cycle++

			if x+(40*int(math.Floor((float64(cycle)/40)))) <= cycle && cycle <= x+2+(40*int(math.Floor((float64(cycle)/40)))) {
				pixels[cycle-1] = "#"
			}

			if printing {
				fmt.Print("During cycle  ", cycle, ": CRT draws pixel in position ", cycle-1, "\n")
				fmt.Print("Current CRT row: ", pixels[:cycle], "\n")
			}

			xp, _ := strconv.Atoi((strings.TrimSpace(instruction[5:])))
			x += xp
		}
		if printing {
			fmt.Print("End of cycle  ", cycle, ": finish executing ", instruction[:len(instruction)-1], " (Register X is now ", x, ")\n")
		}
		cycle++
	}

	// Print the CRT
	for row := 0; row < 6; row++ {
		fmt.Print(pixels[0+(40*row):40+(40*row)], "\n")
	}

	return 0
}

func main() {

	testfile := "tests.txt"
	inputfile := "inputs.txt"

	fmt.Println("Advent-Of-Code 2022 - Day10")
	fmt.Printf("Tests : Answer to Part 1 = %v\n", Part1(testfile))
	fmt.Printf("Inputs: Answer to Part 1 = %v\n", Part1(inputfile))
	fmt.Printf("Tests : Answer to Part 2 = %v\n", Part2(testfile, false))
	fmt.Printf("Inputs: Answer to Part 2 = %v\n", Part2(inputfile, false))

}
